lib/gooddata/core/splunk_logger.rb

```diff
@@ -0,0 +1,201 @@
+          @modes[mode] = modes[mode] if modes.key?(mode)
+        end
+      end
+
+      @rest_aggregator = GoodData::Rest::Aggregator.new
+
+      @context = {
+        :api_version => GoodData.version,
+        :log_v => 0,
+        :action => "undefined",
+        :brick => "undefined",
+        :status => "outside"
+      }
+      @action_start = Time.now
+
+      @params_filter = nil
+
+      @formatter = proc do |severity, datetime, progname, msg|
+        filter_string! msg unless msg.nil?
+        filter_string! progname unless progname.nil?
+        @default_formatter.call severity, datetime, progname, msg
+      end
+    end
+
+    def filter_string!(str)
+      str.gsub! @params_filter, "****" unless @params_filter.nil?
+    end
+
+    def params_filter(map)
+      @params_filter = filter_from_param_array map.values
+      @params_filter = Regexp.new @params_filter.join("|")
+    end
+
+    def filter_from_param_array(arr)
+      filter = []
+      arr.each do |val|
+        filter << "(" + Regexp.quote(val) + ")" if val.class == String
+        filter.concat filter_from_param_array(val) if val.class == Array
+        filter.concat filter_from_param_array(val.values) if val.class == Hash
+      end
+      filter
+    end
+
+    def mode_on(mode)
+      @modes[mode] = true
+    end
+
+    def mode_off(mode)
+      @modes[mode] = false
+    end
+
+    [:buffering, :api_output, :file_output].each do |mode|
+      define_method :"#{mode}_on" do
+        mode_on(mode)
+      end
+
+      define_method :"#{mode}_off" do
+        mode_off(mode)
+      end
+
+      define_method :"#{mode}_on?" do
+        @modes[mode]
+      end
+    end
+
+    def time_from_action_start(t)
+      (t - @action_start) * 1000
+    end
+
+    def start_action(action)
+      flush_aggregator
+
+      now = Time.now
+      end_action
+      set_context(ACTION_CONTEXT, action)
+      @action_start = now
+      set_context ACTION_STATUS_CONTEXT, 'start'
+      add INFO, nil, '', now
+      set_context ACTION_STATUS_CONTEXT, 'in_progress'
+    end
+
+    def end_action(now = Time.now)
+      return if @context[:action] == 'undefined' || @context[:status] == 'outside'
+      flush_aggregator
+
+      set_context ACTION_STATUS_CONTEXT, 'end'
+      add INFO, nil, '', now
+      set_context ACTION_STATUS_CONTEXT, 'outside'
+      set_context ACTION_CONTEXT, 'undefined'
+    end
+
+    def set_context(key, val)
+      @context[key] = val
+    end
+
+    def clear_logs
+      @logs = ""
+    end
+
+    def add(severity, message = nil, progname = nil, time = nil)
+      severity ||= UNKNOWN
+      if @logdev.nil? or severity < @level
+        return true
+      end
+      progname ||= @progname
+      if message.nil?
+        if block_given?
+          message = yield
+        else
+          message = progname
+          progname = @progname
+        end
+      end
+
+      now = (time.nil? ? Time.now : time)
+
+      if !progname.nil? && progname.eql?("rest_call")
```

Comment:
Style/IfUnlessModifier: Favor modifier `if` usage when having a single-line body. Another good alternative is the usage of control flow `&&`/`||`.